llvm/lib/Transforms/IPO/GlobalOpt.cpp

```diff
@@ -2498,6 +2499,102 @@ OptimizeGlobalAliases(Module &M,
   return Changed;
 }
 
+static bool AddRangeMetadata(Module &M) {
+  const DataLayout &DL = M.getDataLayout();
+  bool Changed = false;
+
+  for (GlobalValue &Global : M.global_values()) {
```

Comment:
Wonder whether this can be integrated in one of the existing loops over globals?

Reply:
I didn't try to find a place like that because all other transformation of GlobalOpt are iterated on until GlobalOpt can do no changes. and this process only needs to happen once.

Reply:
Okay, that makes sense.